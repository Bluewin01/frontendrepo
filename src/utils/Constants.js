export const theme = {
  color: {
    w: "#FFFFFF",
    k: "#000000",
    p1: "#D31145",
    p2: "#596C80",
    p3: "#554344",
    p4: "#363E3F",
    p5: "#e6edeb",
    s1: "#E86487",
    s2: "#FBBF56",
    s3: "#97CB5D",
    s4: "#46D3BD",
    s5: "#22A8DA",
    s6: "#9962D2",
    t1: "#F6CCD7",
    t2: "#B8123E",
    t3: "#F9E1B8",
    t4: "#C36418",
    t5: "#C6EB9E",
    t6: "#65962E",
    b1: "#D4EDF1",
    b2: "#FAF9F3",
    b3: "#F0EEE1",
    b4: "#DBDAD4",
    b5: "#B3B1A8",
    b6: "#6F6E68",
    b8: "#302F2C",
    b9: "#292025",
    u2: "#FF5500",
    u3: "#405B70",
    u4: "#769FAB",
  },
};

export const BASE_API = `${process.env.REACT_APP_BASE_API_URL}`;
