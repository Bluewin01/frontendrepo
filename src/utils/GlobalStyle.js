import { createGlobalStyle } from 'styled-components';
import 'antd/dist/antd.css';
import '@aia-digital/ui-library/lib/static/styles.css';

import AIABoldEOT from '../assets/fonts/AIAEverest-Bold.eot';
import AIABoldWOFF from '../assets/fonts/AIAEverest-Bold.woff';
import AIABoldWOFF2 from '../assets/fonts/AIAEverest-Bold.woff2';
import AIABoldTTF from '../assets/fonts/AIAEverest-Bold.ttf';
import AIABoldSVG from '../assets/fonts/AIAEverest-Bold.svg';

import AIACondensedEOT from '../assets/fonts/AIAEverest-Condensed.eot';
import AIACondensedWOFF from '../assets/fonts/AIAEverest-Condensed.woff';
import AIACondensedWOFF2 from '../assets/fonts/AIAEverest-Condensed.woff2';
import AIACondensedTTF from '../assets/fonts/AIAEverest-Condensed.ttf';
import AIACondensedSVG from '../assets/fonts/AIAEverest-Condensed.svg';

import AIACondensedMediumEOT from '../assets/fonts/AIAEverest-CondensedMedium.eot';
import AIACondensedMediumWOFF from '../assets/fonts/AIAEverest-CondensedMedium.woff';
import AIACondensedMediumWOFF2 from '../assets/fonts/AIAEverest-CondensedMedium.woff2';
import AIACondensedMediumTTF from '../assets/fonts/AIAEverest-CondensedMedium.ttf';
import AIACondensedMediumSVG from '../assets/fonts/AIAEverest-CondensedMedium.svg';

import AIAExtraBoldEOT from '../assets/fonts/AIAEverest-ExtraBold.eot';
import AIAExtraBoldWOFF from '../assets/fonts/AIAEverest-ExtraBold.woff';
import AIAExtraBoldWOFF2 from '../assets/fonts/AIAEverest-ExtraBold.woff2';
import AIAExtraBoldTTF from '../assets/fonts/AIAEverest-ExtraBold.ttf';
import AIAExtraBoldSVG from '../assets/fonts/AIAEverest-ExtraBold.svg';

import AIAMediumEOT from '../assets/fonts/AIAEverest-Medium.eot';
import AIAMediumWOFF from '../assets/fonts/AIAEverest-Medium.woff';
import AIAMediumWOFF2 from '../assets/fonts/AIAEverest-Medium.woff2';
import AIAMediumTTF from '../assets/fonts/AIAEverest-Medium.ttf';
import AIAMediumSVG from '../assets/fonts/AIAEverest-Medium.svg';

import AIARegularEOT from '../assets/fonts/AIAEverest-Regular.eot';
import AIARegularWOFF from '../assets/fonts/AIAEverest-Regular.woff';
import AIARegularWOFF2 from '../assets/fonts/AIAEverest-Regular.woff2';
import AIARegularTTF from '../assets/fonts/AIAEverest-Regular.ttf';
import AIARegularSVG from '../assets/fonts/AIAEverest-Regular.svg';

const GlobalStyle = createGlobalStyle`
  @font-face {
    font-family: 'AIABold';
    src: url('${AIABoldEOT}');
    src: url('${AIABoldEOT}?#iefix') format('embedded-opentype'),
        url('${AIABoldWOFF}') format('woff'),
        url('${AIABoldWOFF2}') format('woff2'),
        url('${AIABoldTTF}') format('truetype'),
        url('${AIABoldSVG}#AIAEverestBold') format('svg');
    font-weight: normal;
    font-style: normal;
  }
  @font-face {
    font-family: 'AIACondensed';
    src: url('${AIACondensedEOT}');
    src: url('${AIACondensedEOT}?#iefix') format('embedded-opentype'),
        url('${AIACondensedWOFF}') format('woff'),
        url('${AIACondensedWOFF2}') format('woff2'),
        url('${AIACondensedTTF}') format('truetype'),
        url('${AIACondensedSVG}#AIAEverestCondensed') format('svg');
    font-weight: normal;
    font-style: normal;
  }
  @font-face {
    font-family: 'AIACondensedMedium';
    src: url('${AIACondensedMediumEOT}');
    src: url('${AIACondensedMediumEOT}?#iefix') format('embedded-opentype'),
        url('${AIACondensedMediumWOFF}') format('woff'),
        url('${AIACondensedMediumWOFF2}') format('woff2'),
        url('${AIACondensedMediumTTF}') format('truetype'),
        url('${AIACondensedMediumSVG}#AIAEverestCondensedMedium') format('svg');
    font-weight: 500;
    font-style: normal;
  }
  @font-face {
    font-family: 'AIAExtraBold';
    src: url('${AIAExtraBoldEOT}');
    src: url('${AIAExtraBoldEOT}?#iefix') format('embedded-opentype'),
        url('${AIAExtraBoldWOFF}') format('woff'),
        url('${AIAExtraBoldWOFF2}') format('woff2'),
        url('${AIAExtraBoldTTF}') format('truetype'),
        url('${AIAExtraBoldSVG}#AIAEverestExtraBold') format('svg');
    font-weight: normal;
    font-style: normal;
  }
  @font-face {
    font-family: 'AIAMedium';
    src: url('${AIAMediumEOT}');
    src: url('${AIAMediumEOT}?#iefix') format('embedded-opentype'),
        url('${AIAMediumWOFF}') format('woff'),
        url('${AIAMediumWOFF2}') format('woff2'),
        url('${AIAMediumTTF}') format('truetype'),
        url('${AIAMediumSVG}#AIAEverestMedium') format('svg');
    font-weight: 600;
    font-style: normal;
  }
  @font-face {
    font-family: 'AIARegular';
    src: url('${AIARegularEOT}');
    src: url('${AIARegularEOT}?#iefix') format('embedded-opentype'),
        url('${AIARegularWOFF}') format('woff'),
        url('${AIARegularWOFF2}') format('woff2'),
        url('${AIARegularTTF}') format('truetype'),
        url('${AIARegularSVG}#AIAEverestRegular') format('svg');
    font-weight: 400;
    font-style: normal;
  }

  body h1 {
    font-family: "AIACondensedMedium", "Calibri", sans-serif;
    font-size: 96px;
    line-height: 84px;
    font-weight: normal;
    margin-top: 0px;
    margin-bottom: 0px;
    color: ${({ theme }) => theme.color.p3}
  }

  body h2 {
    font-family: "AIACondensedMedium", "Calibri", sans-serif;
    font-size: 72px;
    line-height: 64px;
    font-weight: normal;
    margin-top: 0px;
    margin-bottom: 0px;
    color: ${({ theme }) => theme.color.p3}
  }

  body h3 {
    font-family: "AIACondensedMedium", "Calibri", sans-serif;
    font-size: 48px;
    line-height: 44px;
    font-weight: normal;
    margin-top: 0px;
    margin-bottom: 0px;
    color: ${({ theme }) => theme.color.p3}
  }

  body h4 {
    font-family: "AIACondensedMedium", "Calibri", sans-serif;
    font-size: 32px;
    line-height: 28px;
    font-weight: normal;
    margin-top: 0px;
    margin-bottom: 0px;
    color: ${({ theme }) => theme.color.p3}
  }

  body h5 {
    font-family: "AIACondensedMedium", "Calibri", sans-serif;
    font-size: 24px;
    line-height: 22px;
    font-weight: normal;
    margin-top: 0px;
    margin-bottom: 0px;
    color: ${({ theme }) => theme.color.p3}
  }

  body h6 {
    font-family: "AIACondensedMedium", "Calibri", sans-serif;
    font-size: 20px;
    line-height: 20px;
    font-weight: normal;
    margin-top: 0px;
    margin-bottom: 0px;
    color: ${({ theme }) => theme.color.p3}
  }

  body p, ul, ol {
    font-family: "AIARegular", Arial, "Helvetica Neue", Helvetica, sans-serif;
    font-size: 15px;
    line-height: 20px;
    margin-bottom: 0px;
    color: ${({ theme }) => theme.color.p3}
  }
`;

export default GlobalStyle;
