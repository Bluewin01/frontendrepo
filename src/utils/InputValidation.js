export const validateIsInputEmpty = (props) => {
  const inputValues = Object.values(props);
  const inputKeys = Object.keys(props);
  let invalidInputList = [];

  inputValues.forEach((input, index) => {
    if (
      input === "" ||
      input === undefined ||
      input === null ||
      input.match(/^ *$/) !== null
    ) {
      invalidInputList.push(index);
    }
  });

  let validationResult = [];
  invalidInputList.forEach((fieldIndex) => {
    validationResult.push(inputKeys[fieldIndex]);
  });

  return validationResult;
};

export const validateIsInputListEmpty = (props) => {
  let invalidInputList = [];

  props.forEach((input) => {
    if (
      input.value === "" ||
      input.value === undefined ||
      input.value === null ||
      input.value.match(/^ *$/) !== null
    ) {
      invalidInputList.push(input.name);
    }
  });

  return invalidInputList;
};

export const validateEmailFormat = (props) => {
  const inputValues = Object.values(props);
  const inputKeys = Object.keys(props);
  let invalidEmailList = [];

  inputValues.forEach((email, index) => {
    if (
      !email.match(
        /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
      )
    ) {
      invalidEmailList.push(index);
    }
  });

  let validationResult = [];
  invalidEmailList.forEach((fieldIndex) => {
    validationResult.push(inputKeys[fieldIndex]);
  });

  return validationResult;
};

export const validateEmailListFormat = (props) => {
  let invalidEmailList = [];

  props.forEach((email) => {
    if (email.value !== "") {
      if (
        !email.value.match(
          /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
        )
      ) {
        invalidEmailList.push(email.name);
      }
    }
  });

  return invalidEmailList;
};
