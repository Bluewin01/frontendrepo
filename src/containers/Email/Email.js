import { connect } from "react-redux";
import EmailEditor from "react-email-editor";
import React, { useRef, useEffect, useState } from "react";
import { Modal } from "../../components/Modal/Modal";
import * as TemplateActions from "../../store/Template/TemplateActions";
import PropTypes from "prop-types";
import { ToastContainer, toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import Loader from "../../components/Loader/Loader";
import "../../assets/css/editor.css";
import History from "../../routes/History";
import { createStructuredSelector } from "reselect";
import { makeSelectJwt } from "../../store/Auth/AuthReselect";
import {
  makeSelectResponse,
  makeSelectError,
  makeSelectTemplateObject,
  makeSelectReady,
} from "../../store/Template/TemplateReselect";
import {
  validateIsInputEmpty,
  validateIsInputListEmpty,
  validateEmailListFormat,
  validateEmailFormat,
} from "../../utils/InputValidation";
import { Button } from "react-bootstrap";
import { InputField } from "../../components/InputField/InputField";
import { MultiInputField } from "../../components/MultiInputField/MultiInputField";

const Email = (props) => {
  const {
    initiateSaveTemplate,
    initiateSendEmail,
    error,
    response,
    initiateSetResponseToDefault,
    initiateSetErrorToDefault,
    initiateEditTemplate,
    templateObject,
    ready,
    initiateSetReady,
  } = props;

  const [loading, showLoader] = useState(false);
  const emailEditorRef = useRef(null);
  const [name, setName] = useState("");
  const [subject, setSubject] = useState("");
  const [description, setDescription] = useState("");
  const [from, setFrom] = useState("");
  const [fromError, setFromError] = useState("");
  const [nameError, setNameError] = useState("");
  const [subjectError, setSubjectError] = useState("");
  const [descriptionError, setDescriptionError] = useState("");
  const [showSaveModal, setShowSaveModal] = useState(false);
  const [showCancelModal, setShowCancelModal] = useState(false);
  const [showTestModal, setShowTestModal] = useState(false);
  const [toList, setToList] = useState([]);
  const [ccList, setCcList] = useState([]);
  const [bccList, setBccList] = useState([]);
  const [toCounter, setToCounter] = useState(1);
  const [ccCounter, setCcCounter] = useState(1);
  const [bccCounter, setbccCounter] = useState(1);

  const buttons = [
    {
      name: "Test",
      color: "danger",
      handler: () => {
        setShowTestModal(true);
      },
    },
    {
      name: "Save",
      color: "danger",
      handler: () => {
        setShowSaveModal(true);
      },
    },
    {
      name: "Cancel",
      color: "light",
      handler: () => {
        setShowCancelModal(true);
      },
    },
  ];

  const initiateErrorMessage = () => {
    setFromError("");
    setSubjectError("");
    setDescriptionError("");
    setFromError("");
    setToList(() => {
      toList.forEach((email) => {
        email.error = "";
      });
      return [...toList];
    });
  };

  useEffect(() => {
    if (templateObject !== null) {
      const { name, subject, description } = templateObject;
      setName(name);
      setSubject(subject);
      setDescription(description);
      showLoader(true);
      setTimeout(() => {
        showLoader(false);
        onLoad(templateObject);
      }, 3000);
    } else {
      setName("");
      setSubject("");
      setDescription("");
    }
  }, [templateObject]);

  useEffect(() => {
    if (error) toast.error(error);
    if (response) toast.success(response);
    initiateSetErrorToDefault();
    initiateSetResponseToDefault();
  }, [error, response]);

  useEffect(() => {
    setToList([
      {
        error: "",
        name: `To0`,
        value: "",
        placeholder: "e.g: email@aia.com",
        required: true,
        setInput: setToList,
      },
    ]);

    setCcList([
      {
        error: "",
        name: `Cc0`,
        value: "",
        placeholder: "e.g: email@aia.com",
        setInput: setCcList,
      },
    ]);

    setBccList([
      {
        error: "",
        name: `Bcc0`,
        value: "",
        placeholder: "e.g: email@aia.com",
        setInput: setBccList,
      },
    ]);
  }, []);

  const handleSave = () => {
    initiateErrorMessage();
    const validationResult = validateIsInputEmpty({
      name,
      subject,
      description,
    });
    if (validationResult.length === 0) {
      emailEditorRef.current.editor.exportHtml(
        (data) => {
          const type = "Email";
          const { design, html } = data;
          const rawValue = design;
          const value = html.replace(/^\s+|\s+$|\n/gm, "").replace(/"/gm, "'");
          if (templateObject !== null) {
            initiateEditTemplate(
              name,
              subject,
              description,
              value,
              type,
              rawValue
            );
          } else {
            initiateSaveTemplate(
              name,
              subject,
              description,
              value,
              type,
              rawValue
            );
          }
          setTimeout(() => {
            History.push("/home");
          }, 1000);
        },
        {
          cleanup: true,
        }
      );
    } else {
      if (validationResult.includes("name")) setNameError("Name is required.");
      if (validationResult.includes("subject"))
        setSubjectError("Subject is required.");
      if (validationResult.includes("description"))
        setDescriptionError("Desciprtion is required.");
    }
  };

  const handleCancel = () => {
    History.push("/home");
  };

  const handleSendEmail = () => {
    initiateErrorMessage();
    const emptyValidation = validateIsInputEmpty({ from, subject });
    const emptyListValidation = validateIsInputListEmpty(toList);
    const formatValidation = validateEmailFormat({ from });
    const toListFormatValidation = validateEmailListFormat(toList);
    const ccListFormatvalidation = validateEmailListFormat(ccList);
    const bccListFormatvalidation = validateEmailListFormat(bccList);

    if (
      emptyValidation.length === 0 &&
      emptyListValidation.length === 0 &&
      formatValidation.length === 0
    ) {
      var to = [];
      var cc = [];
      var bcc = [];

      toList.forEach((list) => {
        to.push(list.value);
      });
      ccList.forEach((list) => {
        cc.push(list.value);
      });
      bccList.forEach((list) => {
        bcc.push(list.value);
      });

      emailEditorRef.current.editor.exportHtml(
        (data) => {
          const { html } = data;
          const value = html.replace(/^\s+|\s+$|\n/gm, "").replace(/"/gm, "'");
          initiateSendEmail(to, from, value, subject, cc, bcc);
        },
        {
          cleanup: true,
        }
      );
    } else {
      if (emptyListValidation.length > 0) {
        setToList(() => {
          emptyListValidation.forEach((fieldName) => {
            let foundIndex = toList.findIndex(
              (list) => list.name === fieldName
            );
            toList[foundIndex].error = "To is required.";
          });
          return [...toList];
        });
      }
      if (toListFormatValidation.length > 0) {
        setToList(() => {
          toListFormatValidation.forEach((fieldName) => {
            let foundIndex = toList.findIndex(
              (list) => list.name === fieldName
            );
            toList[foundIndex].error = "Invalid email format.";
          });
          return [...toList];
        });
      }
      if (ccListFormatvalidation.length > 0) {
        setCcList(() => {
          ccListFormatvalidation.forEach((fieldName) => {
            let foundIndex = ccList.findIndex(
              (list) => list.name === fieldName
            );
            ccList[foundIndex].error = "Invalid email format.";
          });
          return [...ccList];
        });
      }
      if (bccListFormatvalidation.length > 0) {
        setBccList(() => {
          bccListFormatvalidation.forEach((fieldName) => {
            let foundIndex = bccList.findIndex(
              (list) => list.name === fieldName
            );
            bccList[foundIndex].error = "Invalid email format";
          });
          return [...bccList];
        });
      }

      if (emptyValidation.includes("from")) {
        setFromError("From is required.");
      } else if (formatValidation.includes("from")) {
        setFromError("Invalid email format.");
      }
      if (emptyValidation.includes("subject"))
        setSubjectError("Subject is required.");
    }
  };

  const handleCreateEmailFields = () => {
    setToList([
      ...toList,
      {
        error: "",
        name: `To${toCounter}`,
        value: "",
        placeholder: "e.g: email@aia.com",
        setInput: setToList,
      },
    ]);
    setToCounter(toCounter + 1);
  };

  const handleCreateCcFields = () => {
    setCcList([
      ...ccList,
      {
        error: "",
        name: `To${ccCounter}`,
        value: "",
        placeholder: "e.g: email@aia.com",
        setInput: setCcList,
      },
    ]);
    setCcCounter(ccCounter + 1);
  };

  const handleCreateBccFields = () => {
    setBccList([
      ...bccList,
      {
        error: "",
        name: `To${bccCounter}`,
        value: "",
        placeholder: "e.g: email@aia.com",
        setInput: setBccList,
      },
    ]);
    setbccCounter(bccCounter + 1);
  };

  const handleDeleteEmailFields = (fieldName) => {
    setToList(toList.filter((field) => field.name !== fieldName));
  };

  const handleDeleteCcFields = (fieldName) => {
    setCcList(ccList.filter((field) => field.name !== fieldName));
  };

  const handleDeleteBccFields = (fieldName) => {
    setBccList(bccList.filter((field) => field.name !== fieldName));
  };

  window.onbeforeunload = function () {
    return "Refresh will make you lose your changes";
  };

  const onLoad = (templateObject) => {
    const { rawValue } = templateObject.rawValue;
    if (emailEditorRef.current.editor) {
      emailEditorRef.current.editor.loadDesign(rawValue);
    }
  };

  const onReady = () => {
    initiateSetReady();
  };

  const nameField = {
    value: name,
    error: nameError,
    name: "Name",
    type: "text",
    placeholder: "Enter template name here",
    required: true,
    setInput: setName,
    setError: setNameError,
  };

  const subjectField = {
    value: subject,
    error: subjectError,
    name: "Subject",
    type: "text",
    placeholder: "Enter subject here",
    required: true,
    setInput: setSubject,
    setError: setSubjectError,
  };

  const descriptionField = {
    value: description,
    error: descriptionError,
    name: "Description",
    type: "text",
    placeholder: "Enter description here",
    required: true,
    setInput: setDescription,
    setError: setDescriptionError,
  };

  const toField = {
    label: "To",
    required: true,
    inputFields: toList,
    createField: handleCreateEmailFields,
    deleteField: handleDeleteEmailFields,
  };

  const fromField = {
    value: from,
    error: fromError,
    name: "From",
    type: "text",
    placeholder: "e.g: email@aia.com",
    required: true,
    error: fromError,
    setInput: setFrom,
    setError: setFromError,
  };

  const ccField = {
    label: "Cc",
    inputFields: ccList,
    createField: handleCreateCcFields,
    deleteField: handleDeleteCcFields,
  };

  const bccField = {
    label: "Bcc",
    inputFields: bccList,
    createField: handleCreateBccFields,
    deleteField: handleDeleteBccFields,
  };

  return (
    <div className="text-center">
      <ToastContainer
        position="top-right"
        autoClose={5000}
        hideProgressBar
        newestOnTop={false}
        closeOnClick
        rtl={false}
        pauseOnFocusLoss
        draggable
        pauseOnHover={false}
      />
      {loading ? <Loader></Loader> : null}
      <EmailEditor ref={emailEditorRef} onReady={onReady} />
      <div className="input__form">
        {ready
          ? buttons.map((button) => {
              return (
                <Button
                  key={button.name}
                  className="Button"
                  variant={button.color}
                  onClick={button.handler}
                >
                  {button.name}
                </Button>
              );
            })
          : null}
        {showTestModal && (
          <Modal
            headerTitle={`Test Email Template`}
            buttons={[
              {
                name: "Send Email",
                color: "danger",
                handler: handleSendEmail,
              },
              {
                name: "Cancel",
                color: "light",
                handler: () => setShowTestModal(false),
              },
            ]}
          >
            <MultiInputField {...toField} />
            <InputField {...fromField} />
            <InputField {...subjectField} />
            <MultiInputField {...ccField} />
            <MultiInputField {...bccField} />
          </Modal>
        )}
        {showSaveModal && (
          <Modal
            headerTitle={`Are you sure you want to save Email template?`}
            buttons={[
              {
                name: "Confirm",
                color: "danger",
                handler: handleSave,
              },
              {
                name: "Cancel",
                color: "light",
                handler: () => setShowSaveModal(false),
              },
            ]}
          >
            <InputField {...nameField} />
            <InputField {...subjectField} />
            <InputField {...descriptionField} />
          </Modal>
        )}
        {showCancelModal && (
          <Modal
            headerTitle={`All changes will be lost. Are you sure to continue?`}
            buttons={[
              {
                name: "Confirm",
                color: "danger",
                handler: handleCancel,
              },
              {
                name: "Cancel",
                color: "light",
                handler: () => setShowCancelModal(false),
              },
            ]}
          />
        )}
      </div>
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  error: makeSelectError(),
  response: makeSelectResponse(),
  jwt: makeSelectJwt(),
  templateObject: makeSelectTemplateObject(),
  ready: makeSelectReady(),
});

const mapDispatchToProps = (dispatch) => {
  return {
    initiateSaveTemplate: (name, subject, description, value, type, rawValue) =>
      dispatch(
        TemplateActions.saveTemplateRequest({
          name,
          subject,
          description,
          value,
          type,
          rawValue,
        })
      ),
    initiateSendEmail: (to, from, value, subject, cc, bcc) =>
      dispatch(
        TemplateActions.sendEmailRequest({
          to,
          from,
          value,
          subject,
          cc,
          bcc,
        })
      ),
    initiateSetResponseToDefault: () =>
      dispatch(TemplateActions.setResponseToDefault()),
    initiateSetErrorToDefault: () =>
      dispatch(TemplateActions.setErrorToDefault()),
    initiateEditTemplate: (
      name,
      subject,
      description,
      value,
      type,
      rawValue,
      userCode
    ) =>
      dispatch(
        TemplateActions.editTemplateRequest({
          name,
          subject,
          description,
          value,
          type,
          rawValue,
          userCode,
        })
      ),
    initiateSetReady: () => dispatch(TemplateActions.setReady()),
  };
};

Email.propTypes = {
  initiateSaveTemplate: PropTypes.func,
  initiateSendEmail: PropTypes.func,
  response: PropTypes.string,
  error: PropTypes.string,
  initiateSetResponseToDefault: PropTypes.func,
  initiateSetErrorToDefault: PropTypes.func,
  initiateEditTemplate: PropTypes.func,
  templateObject: PropTypes.object,
  initiateSetReady: PropTypes.func,
};

export default connect(mapStateToProps, mapDispatchToProps)(Email);
