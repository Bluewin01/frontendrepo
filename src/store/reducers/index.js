import { combineReducers } from "redux";
import { configureStore } from "@reduxjs/toolkit";
import lang from "../Lang/LangReducer";
import loader from "../Loader/LoaderReducer";
import Auth from "../Auth/AuthReducer";
import Template from "../Template/TemplateReducer";

const rootReducer = combineReducers({
  lang,
  loader,
  Auth,
  Template,
});

export const setupStore = (preloadedState) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export default rootReducer;
