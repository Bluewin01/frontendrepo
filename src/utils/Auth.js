// Service to check authentication for user and to signOut
import * as AuthActions from "../store/Auth/AuthActions";
import store from "../store";

const Auth = {
  signOut() {
    store.dispatch(AuthActions.signOut);
  },
  isAuth(jwt) {
    if (!jwt) {
      return false;
    } else {
      return true;
    }
  },
};
export default Auth;
