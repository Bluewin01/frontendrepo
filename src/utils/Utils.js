// test-utils.jsx
import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { setupStore } from "../store/reducers";

export function render(
  ui,
  { preloadedState, store = setupStore(preloadedState), ...renderOptions } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from "@testing-library/react";
