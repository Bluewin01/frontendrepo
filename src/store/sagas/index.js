import { AuthSaga } from "../Auth/AuthSaga";
import { TemplateSaga } from "../Template/TemplateSaga";
import { fork, all } from "redux-saga/effects";

export function* watchSagas() {
  yield all([AuthSaga(), TemplateSaga()]);
}
