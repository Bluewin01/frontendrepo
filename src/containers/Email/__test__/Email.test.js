import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { screen, cleanup, fireEvent } from "@testing-library/react";
import { render } from "../../../utils/Utils";
import renderer from "react-test-renderer";
import store from "../../../store";
import { Provider } from "react-redux";
import Email from "../Email";

jest.useFakeTimers();

describe("Containers: Email Editor", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });
  afterEach(cleanup);

  describe("Positive Case", () => {
    it("should be render correctly", () => {
      const tree = renderer
        .create(
          <Provider store={store}>
            <Email />
          </Provider>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("should be able to render email page with error", () => {
      render(<Email />, {
        preloadedState: {
          Template: {
            ready: true,
            error: "fakeError",
            response: null,
            templateObject: null,
          },
        },
      });
    });

    it("should be able to render email page with response", () => {
      render(<Email />, {
        preloadedState: {
          Template: {
            ready: true,
            error: null,
            response: "fakeResponse",
            templateObject: null,
          },
        },
      });
    });

    it("should be shown confirm & cancel button when click on save button", async () => {
      render(<Email />, {
        preloadedState: {
          Template: {
            ready: true,
            templateObject: null,
          },
        },
      });

      const saveButton = screen.getByRole("button", { name: "Save" });

      expect(saveButton).toBeInTheDocument();

      fireEvent.click(saveButton);

      const confirmButton = screen.getByRole("button", { name: "Confirm" });
      const cancelButton = screen.getByRole("button", { name: "Cancel" });

      expect(confirmButton).toBeInTheDocument();
      expect(cancelButton).toBeInTheDocument();

      fireEvent.click(confirmButton);
    });

    it("should be shown send email & cancel button when click on test button", async () => {
      render(<Email />, {
        preloadedState: {
          Template: {
            ready: true,
            templateObject: null,
          },
        },
      });

      const testButton = screen.getByRole("button", { name: "Test" });

      expect(testButton).toBeInTheDocument();

      fireEvent.click(testButton);

      const sendEmailButton = screen.getByRole("button", {
        name: "Send Email",
      });
      const cancelButton = screen.getByRole("button", { name: "Cancel" });

      expect(sendEmailButton).toBeInTheDocument();
      expect(cancelButton).toBeInTheDocument();
    });

    it("should shown All changes will be lost. Are you sure to continue? when click on cancel button", async () => {
      render(<Email />, {
        preloadedState: {
          Template: {
            ready: true,
            templateObject: null,
          },
        },
      });

      const cancelButton = screen.getByRole("button", { name: "Cancel" });

      expect(cancelButton).toBeInTheDocument();

      fireEvent.click(cancelButton);

      expect(
        screen.getByText("All changes will be lost. Are you sure to continue?")
      ).toBeInTheDocument();
    });

    it("should redirect to home page when click on confirm button", async () => {
      render(<Email />, {
        preloadedState: {
          Template: {
            ready: true,
            templateObject: null,
          },
        },
      });

      const cancelButton = screen.getByRole("button", { name: "Cancel" });

      expect(cancelButton).toBeInTheDocument();

      fireEvent.click(cancelButton);

      const confirmButton = screen.getByRole("button", { name: "Confirm" });

      fireEvent.click(confirmButton);

      expect(global.window.location.href).toContain("/home");
    });

    it("should be shown confirm button when click on save button with passing templateObject", async () => {
      render(<Email />, {
        preloadedState: {
          Template: {
            ready: true,
            templateObject: {
              name: "fakeName",
              subject: "fakeSubject",
              description: "fakeDescription",
            },
          },
        },
      });

      const saveButton = screen.getByRole("button", { name: "Save" });

      fireEvent.click(saveButton);

      const confirmButton = screen.getByRole("button", { name: "Confirm" });

      expect(confirmButton).toBeInTheDocument();
    });
  });

  describe("Negative Case", () => {
    it("should not shown confirm & cancel button when click on cancel button", async () => {
      render(<Email />, {
        preloadedState: {
          Template: {
            ready: true,
            templateObject: null,
          },
        },
      });

      const saveButton = screen.getByRole("button", { name: "Save" });

      expect(saveButton).toBeInTheDocument();

      fireEvent.click(saveButton);

      const confirmButton = screen.getByRole("button", { name: "Confirm" });
      const cancelButton = screen.getByRole("button", { name: "Cancel" });
      fireEvent.click(cancelButton);
      expect(confirmButton).not.toBeInTheDocument();
      expect(cancelButton).not.toBeInTheDocument();
    });

    it("should not shown send email & cancel button when click on cancel button", async () => {
      render(<Email />, {
        preloadedState: {
          Template: {
            ready: true,
            templateObject: null,
          },
        },
      });

      const testButton = screen.getByRole("button", { name: "Test" });

      expect(testButton).toBeInTheDocument();

      fireEvent.click(testButton);

      const sendEmailButton = screen.getByRole("button", {
        name: "Send Email",
      });
      const cancelButton = screen.getByRole("button", { name: "Cancel" });
      fireEvent.click(cancelButton);

      expect(sendEmailButton).not.toBeInTheDocument();
      expect(cancelButton).not.toBeInTheDocument();
    });

    it("should not be shown All changes will be lost. Are you sure to continue? when click on cancel button", async () => {
      render(<Email />, {
        preloadedState: {
          Template: {
            ready: true,
            templateObject: null,
          },
        },
      });

      const cancelButton = screen.getByRole("button", { name: "Cancel" });

      expect(cancelButton).toBeInTheDocument();

      fireEvent.click(cancelButton);

      const cancelButton2 = screen.getByRole("button", { name: "Cancel" });
      const screenText = screen.getByText(
        "All changes will be lost. Are you sure to continue?"
      );

      fireEvent.click(cancelButton2);

      expect(screenText).not.toBeInTheDocument();
    });
  });
});
