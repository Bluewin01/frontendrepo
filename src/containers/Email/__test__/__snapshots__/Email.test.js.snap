// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers: Email Editor Positive Case should be render correctly 1`] = `
<div
  className="text-center"
>
  <div
    className="Toastify"
  />
  <div
    style={
      Object {
        "display": "flex",
        "flex": 1,
        "minHeight": 500,
      }
    }
  >
    <div
      id="editor-1"
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </div>
  <div
    className="input__form"
  />
</div>
`;
